pkg/operator/sync.go

```diff
@@ -744,10 +753,33 @@ func (optr *Operator) syncMachineConfigNodes(_ *renderConfig) error {
 			return err
 		}
 		p := mcoResourceRead.ReadMachineConfigNodeV1OrDie(mcsBytes)
-		_, _, err = mcoResourceApply.ApplyMachineConfigNode(optr.client.MachineconfigurationV1alpha1(), p)
+		mcn, _, err := mcoResourceApply.ApplyMachineConfigNode(optr.client.MachineconfigurationV1alpha1(), p)
 		if err != nil {
 			return err
 		}
+		// if this is the first time we are applying the MCN and the node is ready, set the config version probably
+		if mcn.Spec.ConfigVersion.Desired == upgrademonitor.NotYetSet {
+			err = upgrademonitor.GenerateAndApplyMachineConfigNodeSpec(optr.fgAccessor, pool, node, optr.client)
+			if err != nil {
+				klog.Errorf("Error making MCN spec for Update Compatible: %w", err)
+			}
+		}
+
+	}
+	if mcns != nil {
+		for _, mcn := range mcns.Items {
```

Comment:
Although for most clusters, there should be some single digit nodes, we do do scale/perf testing with thousands of nodes. Probably shouldn't do a O(nodes^2) operation every sync since we probably have a lot of MCNs there.

At least, could you make this a populate map - read from map O(nodes) operation?

Reply:
yeah you're right, pushing changes now. I made a map of the nodes and then for each mcn I checked if the entry exists.